### PR TITLE
Fix inconsistency of the 'CanRead' property after disposing of an HTTP content stream

### PR DESCRIPTION
**Pull request by angelobreuer**
Tracking: https://github.com/dotnet/runtime/issues/38527

### CODE REVIEW

**Comment by lindexi** [comment]
@angelobreuer Thank you. But I think it is a break changed. Changes in the behavior of this attribute may make some otherwise normal logic not work properly now.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs [hunk 1]
```diff
@@ -18,7 +18,7 @@ public HttpContentReadStream(HttpConnection connection) : base(connection)
             {
             }
 
-            public sealed override bool CanRead => true;
+            public sealed override bool CanRead => _disposed == 0;
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs:21]: If the goal here is increased consistency, then a similar change should be made in HttpContentWriteStream.CanWrite:
https://github.com/dotnet/runtime/blob/5996c03a44cb118e53bed8d542427d30f2094f40/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentWriteStream.cs#L20
(using `_connection != null` as an indication of not being disposed).

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs:21]: And also in RawConnectionStream. And maybe some others.

As noted in the issue, I'm not convinced these changes are necessary: one shouldn't be using these objects after they're disposed. But I doubt the changes hurt much, so I'm ok with it, as long as the change is made comprehensively in this component, such that it improves rather that hurts consistency.

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=38528) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=38528) before we can accept your contribution.<br/><br/>:x: angelobreuer [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=38528)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=38528) it.</sub>

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs (hunk 1 above)
Comment (angelobreuer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs:21]: Thanks. I added the missing fixes to the PR.

File: src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs [hunk 2]
```diff
@@ -405,7 +405,7 @@ public WasmHttpReadStream(WasmFetchResponse status)
                 _status = status;
             }
 
-            public override bool CanRead => true;
+            public override bool CanRead => _status != null || _reader != null;
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:408]: The Dispose implementation isn't nulling these out, so I'm not sure this is the right condition.
cc: @kjpou1

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs [hunk 3]
```diff
@@ -18,8 +18,8 @@ public RawConnectionStream(HttpConnection connection) : base(connection)
                 if (NetEventSource.IsEnabled) NetEventSource.Info(this);
             }
 
-            public sealed override bool CanRead => true;
-            public sealed override bool CanWrite => true;
+            public sealed override bool CanRead => _connection != null;
+            public sealed override bool CanWrite => _connection != null;
 
             public override int Read(Span<byte> buffer)
             {
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs:25]: Can you please add some tests as well?

File: src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs (hunk 2 above)
Comment (angelobreuer) [reply src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:408]: The `_status` field is nulled out when the stream was read to the end, see: 

https://github.com/dotnet/runtime/blob/f2211b2732f190d0ca136dd542ed3af686b3b326/src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs#L465-L473

Also, a comment hints that the stream was fully read if `_status` and `_reader` are both `null`, see:

https://github.com/dotnet/runtime/blob/f2211b2732f190d0ca136dd542ed3af686b3b326/src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs#L432-L440

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:408]: Sure... but what happens if the response stream is disposed before it's fully read?

Comment (angelobreuer) [reply src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:408]: Then we could change the check to:
```csharp
public override bool CanRead => (_status != null && !_status.IsDisposed) || (_reader != null && !_reader.IsDisposed);
```

But I don't know if the check would be too 'overkill' for this scenario. Maybe just add a `_disposed` flag?
I could not find a convention for this in the coding guidelines.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs (hunk 3 above)
Comment (angelobreuer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs:25]: Sure. I added the tests to the PR.

File: src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs (hunk 2 above)
Comment (alnikola) [reply src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:408]: In my opinion, `(_status != null && !_status.IsDisposed) || (_reader != null && !_reader.IsDisposed)` looks fine. Don't see a reason for adding one more field.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:408]: It's definitely not worth adding a field.  The simplest answer here is exactly what it's already doing `CanRead => true`, and we shouldn't add any significant code or cost to do better.  I'm fine with us making them return a "correct" result if it's trivial.

**Review by stephentoub — APPROVED** [review]
LGTM.  Thanks for working on this.

**Comment by stephentoub** [comment]
Ugh, somehow I invalidated the branch associated with the PR by pushing a rebased set of changes.  I'll open a new PR with your commits.  Thanks.

**Pull request closed without merging**